Alertmanager: Add Role-Based Access Control via reqAction Field

olegpixel opened the pull request:
**What is this feature?**

This PR introduces Role-Based Access Control ([RBAC](https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/)) for the Alertmanager datasource plugin in Grafana. It adds the reqAction field to the existing API routes, specifying the required action for each route based on Grafana's RBAC system.

**Why do we need this feature?**
The current implementation of the Alertmanager plugin lacks Role-Based Access Control, making it difficult to enforce fine-grained permissions for different users. By adding the reqAction field, this PR ensures that requests to specific endpoints respect custom-configured roles, improving security and flexibility.

**Who is this feature for?**
This feature is useful for Grafana administrators and users who need to control access to Alertmanager-related API routes based on role-based permissions.

**Additional Context**
- This change follows the RBAC guidelines outlined in the [Grafana documentation](https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/custom-role-actions-scopes/).
- The implementation is similar to [another PR](https://github.com/grafana/grafana/pull/87208) that introduced reqAction fields for Prometheus plugin.

**Special notes for your reviewer:**

Please check that:
- [v] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

olegpixel committed:
Add reqAction for alert manager routes

stevesg commented on public/app/plugins/datasource/alertmanager/plugin.json:
```diff
@@ -7,39 +7,47 @@
     },
     {
       "method": "GET",
       "path": "api/v1/alerts",
-      "reqRole": "Admin"
+      "reqRole": "Admin",
+      "reqAction": "alert.instances:read"
```
Passing note: For Mimir this is actually the path for reading the Alertmanager configuration, not alerts (confusingly).

stevesg commented:
Hi, thanks for working on this. We need to make a few tweaks to get this merged:

- My comment on the `api/v1/alerts` for Mimir datasources - this is the _configuration_ endpoint. `alert.instances:read` is the wrong permission for this.
- Grafana exposes explicit "proxy" routes for alertmanager access, which has some additional smarts in it, e.g. fixes up the paths based on the implementation ([see here](https://github.com/grafana/grafana/blob/3454030bc2209fd3ec88dcf24d00d6db40af27fa/pkg/services/ngalert/api/lotex_am.go#L21-L45)).
- The RBAC for these routes is applied [here](https://github.com/grafana/grafana/blob/3454030bc2209fd3ec88dcf24d00d6db40af27fa/pkg/services/ngalert/api/authorization.go#L212-L228)

I might be wrong as I'm not an RBAC expert, however I believe the existing `alert.instances.external:write` and ``alert.instances.external:read` permissions already solve your problem?

If the desire is to have more fine-grained RBAC for e.g. silences vs alert instances, then we could consider defining 	`alert.silences.external:read` and `alert.silences.external:write`, that would match the Grafana routes.

yuri-tceretian commented:
+1 to what Steve says. Historically, there was no way to protect a specific endpoint by an action. Now if you want to protect silence API by its own action you have to support both: old and new ways. In our APIs we use `or` semantics to not break users that use the old blanket permissions. I am not sure the plugins support it, though.

olegpixel committed:
Remove reqAction from api/v1/alerts endpoint

olegpixel commented:
Hi @stevesg and @yuri-tceretian,

Thank you both for taking the time to review this PR. I truly appreciate your input and feedback.

To address the points in question:

1. **User Roles in Grafana**: In our current Grafana setup, our users possess the following roles: `alert.instances.external:read`, `alert.instances.external:write`, `alert.silences:read`, `alert.silences:create`, and `alert.silences:write`. These roles should ideally grant the necessary permissions for accessing and managing alerts and silences.

2. **UI Endpoint Call**: Our UI is designed to call the endpoint at https://grafana-url.com/api/datasources/proxy/uid/alertmanager-UID/alertmanager/api/v2/silences, where `alertmanager-UID` corresponds to a Mimir UID. However, we're encountering a `403 Forbidden` error with the message 
`{"message":"plugin proxy route access denied"}`.  This suggests there might be a permissions misalignment or an issue with the plugin/proxy configuration.

3. **Removal of reqAction from `api/v1/alerts` Endpoint**. @stevesg, as you mentioned above, `api/v1/alerts` needs another permission specified. I've removed the `reqAction` from the `api/v1/alerts` endpoint in my PR as:
- We are not currently utilising this particular endpoint in our setup.
- I was unable to locate the appropriate permission documentation for this endpoint. [[Link to documentation](https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/custom-role-actions-scopes/)].

I hope this clarifies the rationale behind the changes. Please feel free to ask if there are any further questions or if additional details are needed.

yuri-tceretian reviewed (changes requested):
Thank you for contributing but this needs some changes before we can merge it.

1. actions like `alert.silences` or `alert.instances` have a different scope. The scope for external actions is `datasource UID` and the RBAC for plugins supports only it. 
2. This file contains general permissions for all paths, e.g. entry without `path` means that it matches all paths . Therefore, entry that matches `POST` method cannot be protected by specific action because it matches config endpoint. 
3. Paths that start with `/alertmanager` are for Mimir implementation, other paths for Prometheus implementation. Therefore, I think we need to create dedicated matches for all paths

Here is the map paths to actions that are supported via external data sources alerting API https://github.com/grafana/grafana/blob/309a2eb4e9c771528e97986c27fe9213cce26e78/pkg/services/ngalert/api/authorization.go#L212-L232

We need to check all paths Alertmanager supports and add corresponding entries. For example from here https://github.com/grafana/prometheus-alertmanager/blob/ce9fd632f569516b691d4e08ef0ba524bc584c48/api/v2/openapi.yaml
+ mirror them for Mimir alertmanager

olegpixel committed:
Update public/app/plugins/datasource/alertmanager/plugin.json

Co-authored-by: Yuri Tseretyan <tseretyan@fastmail.com>

olegpixel committed:
Update public/app/plugins/datasource/alertmanager/plugin.json

Co-authored-by: Yuri Tseretyan <tseretyan@fastmail.com>

olegpixel committed:
Update public/app/plugins/datasource/alertmanager/plugin.json

Co-authored-by: Yuri Tseretyan <tseretyan@fastmail.com>

olegpixel committed:
Update public/app/plugins/datasource/alertmanager/plugin.json

Co-authored-by: Yuri Tseretyan <tseretyan@fastmail.com>

olegpixel committed:
Update public/app/plugins/datasource/alertmanager/plugin.json

Co-authored-by: Yuri Tseretyan <tseretyan@fastmail.com>

olegpixel committed:
Fix reqAction

olegpixel commented:
@yuri-tceretian I've updated the `plugin.json` file to address your comments. Please have a look.

There are two endpoints listed in the [api/v2/openapi.yaml](https://github.com/grafana/prometheus-alertmanager/blob/main/api/v2/openapi.yaml) file:
- `/api/v2/status`
- `/api/v2/receivers`

Should these be added as well? I don't see them in the [api/authorization.go](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/api/authorization.go) file.

yuri-tceretian commented:
@olegpixel yes, please add them. The alerting external Alertmanager proxy, to which the snippet of `autorization.go` belongs, does not expose all paths. In contrast, the data source proxy is transparent, and it does not limit the paths users can send requests to. So `plugin.json` needs to enumerate all supported paths that Alertmanager exposes.

yuri-tceretian commented on public/app/plugins/datasource/alertmanager/plugin.json:
```diff
@@ -33,13 +42,27 @@
+      "reqAction": "alert.instances.external:read"
+    },
+    {
+      "method": "POST",
```
please move POST above wildcard check
```
    {
      "method": "POST",
      "reqRole": "Editor"
    },
```
because I think it will catch the check first

olegpixel committed:
Add endpoints from Prometheus Alertmanager API

olegpixel commented:
@yuri-tceretian Thanks for the prompt check and response. I've added additional endpoints/paths from the Prometheus Alertmanager API. Also, I've moved the wildcard checks to the bottom so that they catch only paths that are not explicitly listed above. 
Let me know if anything else is needed to be addressed.
Thank you 🚀

olegpixel committed:
Fix prettier

yuri-tceretian commented:
@olegpixel , sorry for the delay, I will take a look asap.

MylesBR commented:
Thank you @yuri-tceretian please keep me posted

yuri-tceretian commented on public/app/plugins/datasource/alertmanager/plugin.json:
```diff
@@ -4,42 +4,77 @@
+      "method": "GET",
+      "path": "alertmanager/api/v2/status",
+      "reqRole": "Viewer",
+      "reqAction": "alert.instances.external:read"
```
let's change it to `alert.notifications.external:read` because it returns configuration. So, it's similar to `api/v1/alerts` below

yuri-tceretian commented on public/app/plugins/datasource/alertmanager/plugin.json:
```diff
@@ -4,42 +4,77 @@
       "method": "DELETE",
       "path": "alertmanager/api/v2/silence",
-      "reqRole": "Editor"
+      "reqRole": "Editor",
```
let's add slash at the end of `alertmanager/api/v2/silence` to reduce the scope when `HasPrefix` is positive.

yuri-tceretian reviewed (commented):
Left comments. 
Also, paths that have prefix `alertmanager/` will work for Mimir alertmanager only because it adds this prefix, which actually can be change but I think we should not try to cover this here. 
However, we still need to cover Prometheus Alertmanager APIs. So, let's duplicate all those paths without the prefix. Later we will figure out some better fix for this.

yuri-tceretian committed:
add slash to alertmanager/api/v2/silence

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian committed:
use notifications reader for alertmanager/api/v2/status

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian committed:
use viewer for GET api/v1/alerts

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian committed:
add POST DELETE for api/v1/alerts

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian committed:
duplicate paths to cover upstream Alertmanager

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian committed:
add fallback rule to catch all GET

Signed-off-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

yuri-tceretian commented:
To speed up the review cycle I decided to address all the comments.

yuri-tceretian committed:
Merge remote-tracking branch 'origin/main' into fork/olegpixel/alertmanager-plugin-add-reqActions

yuri-tceretian commented:
@JacobsonMT, can you please do sanity check of the changes?

yuri-tceretian committed:
Update reqAction for GET api/v1/alerts

yuri-tceretian committed:
require Admin for all wildcard paths

JacobsonMT reviewed (approved):
LGTM :rocket:

yuri-tceretian reviewed (approved):
LGTM
